data/virt_autotest/guest_console_monitor.sh

```diff
@@ -0,0 +1,247 @@
+  esac
+done
+
+#FUNCTION PART
+function get_console() {
+
+local vmguest=$1
+
+expect -c "
+set timeout 60
+
+##hide echo
+log_user 0
+spawn -noecho virsh console ${vmguest}
+
+#wait connection
+sleep 3
+send \"\r\n\r\n\r\n\"
+
+#condition expect
+expect {
+        \"*login:\" {
+                send \"root\r\"
+                exp_continue
+        }
+        -nocase \"password:\" {
+                send \"novell\r\"
+                exp_continue
+        }
+        \"*:~ #\" {
+                send -- \"ip route get 1\r\"
+        }
+        timeout {
+                send -- \"exit\r\"
+                exp_continue        
+        }                 
+}
+
+## -1 means never timeout
+set timeout -1
+
+expect -re {dev.*\s([0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})[^0-9]}
+
+if {![info exists expect_out(1,string)]} {
+        puts \"Match did not happen :(\"
+}
+set output \$expect_out(buffer)
+expect eof
+
+puts \"\$output\"
+"
+
+}
+
+function open_console() {
+
+    #open guest console
+    for vmguest in ${vm_guestnames_state_running_array[@]};do
+        get_console ${vmguest} >> ${LOG_DIR}/${vmguest}_console_${timestamp} 2>&1 &
+    done 
+}
+
+function close_console() {
+
+    #close guest console
+    echo -e "Kill PIDs $(basename $0)\n" >> ${monitor_log_file}
+    pid=`ps aux | grep "$(basename $0) -s" | grep -v "grep"| awk '{print $2}'`
+    if [[ -n $pid ]];then
+        kill -9 $pid >> /dev/null 2>&1;sync
+    else
+        echo -e "PIDs $(basename $0) were killed\n" >> ${monitor_log_file}
+    fi
+
+    echo -e "Kill all associated "virsh console" processes\n" >> ${monitor_log_file}
+
+    while :
+    do
+        #wait for checking all associated "virsh console" processes"
+        sleep 3
+
+        #Check all associated "virsh console" processes
+        ps aux | grep "virsh console" | grep -v "grep" >> /dev/null 2>&1
+        if [[ $? -eq 0 ]];then 
+            sub_pid=`ps aux | grep "expect -c" | grep -v "grep"| awk '{print $2}'`
+            if [[ -n $sub_pid ]];then
+                kill -9 $sub_pid >> /dev/null 2>&1;sync
+            fi
+        else
+            break
+        fi
+    done
+}
+
+function end_monitor() {
+
+    #end of monitor for guest console
+    if [ -f ${MONITOR_LOCKED_FILE} ];then
+        echo -e "End of Monitor for guest console:\n" | tee -a ${monitor_log_file}
+        #UNLOCK Monitor for guest console
+        rm -rf ${MONITOR_LOCKED_FILE} >> /dev/null 2>&1
+        #Kill Pids from Monitor for guest console
+        close_console
+    fi
+}
+
+function start_monitor() {
+
+    #Start monitor for guest console
+    ##After boot up all existed vm guest(s) from set_guest_running func
+    ##Figure out vm_guestnames_state_running_array from all existed and boot up vm guest(s) 
+    get_vm_guestnames_state_running=`virsh list --state-running | grep "${vm_guestnames_types}" | awk '{print $2}'`
+    vm_guestnames_state_running_array=$(echo -e ${get_vm_guestnames_state_running})
+
+    while :
```

Comment:
Why is below code block under while, and no break call? How it ends?

Reply:
Why is below code block under while, and no break call? 
```
this while loop will work with MONITOR_LOCKED_FILE together
keep virsh console pid(s) still running in background as purpose. 
```

How it ends?
```
call end_monitor  () to end of this monitor:
set unlock for MONITOR_LOCKED_FILE firstly
and then call close_console ()
```

Reply:
I mean how the while loop ends?

Reply:
start_monitor -> keep virsh console still running 
So, used with this while loop to implement virsh console keep running, thus, there was not the while loop ends.

Reply:
This should be wrong. You already run " get_console ${vmguest} >> ${LOG_DIR}/${vmguest}_console.log 2>&1 &" in background by "&". From code logic, you should ensure the while loop correctly ends up after it finishes its planned function, which is to start console for all guests.

Reply:
No, there was not wrong. 

The current code logic, used with LOCK file and while loop to run " get_console ${vmguest} >> ${LOG_DIR}/${vmguest}_console.log 2>&1 &" in background by "&". 

In the while loop, will continue to make sure vm guest as running, and then make sure if get_console ${vmguest} still ran in background, if not will re-trigger " get_console ${vmguest} >> ${LOG_DIR}/${vmguest}_console.log 2>&1 &" in background by "&".  again 

After start monitor guest console, there were some operation for our deployed guest, so, we do not make sure when to terminate virsh condole guest pid, means that virsh console guest was gone 

as mentioned again .. start_monitor, will keep virsh console guest still running as purpose. 

So, let get_console ${vmguest} ran in the while loop without ends here.

Reply:
had confirmed just only start console for all guests was not enough. 

instance :
After trigger virsh console guest, then try to reboot this guest, this PID of ran virsh console guest will be terminated. 
So, have to re-trigger  virsh console guest to handle guest console again.

Reply:
Ah, I get it now. Sorry for my previous misunderstanding. 

Can you share what drives you to make so intense repeated checking(3 seconds sleep only for next round checking) for "virsh console" processes? Do these console processes constantly got terminated for any reason?

Reply:
Never mind, this script looks too easy, but it will face to the complex situation from our vir functional test. 

Can you share what drives you to make so intense repeated checking(3 seconds sleep only for next round checking) for "virsh console" processes?
```
After start monitor for guest console, we do not known when, what operation for guest from our test, also we do not make sure if these operation will effect(borken) our handle guest console.   
So, the current code logic, sleep 3 sec-> make sure if vm guest still running -> make sure virsh console guest still existed in background.
1. if guest keep running, and virsh console still existed , sleep 3 sec will check them again
2. if guest keep running, and PID of virsh console guest was gone, re-trigger virsh console guest again
3. if guest was not running, sleep 3 sec to check guest status again 
```
  
Do these console processes constantly got terminated for any reason?
`right now, just make sure reboot guest will terminated  guest console`

Reply:
Never mind, this script looks too easy, but it will face to the complex situation from our vir functional test. 

Can you share what drives you to make so intense repeated checking(3 seconds sleep only for next round checking) for "virsh console" processes?
```
After start monitor for guest console, we do not known when, what operation for guest from our test, also we do not make sure if these operation will effect(borken) our handle guest console.   
So, the current code logic, sleep 3 sec-> make sure if vm guest still running -> make sure virsh console guest still existed in background.
1. if guest keep running, and virsh console still existed , sleep 3 sec will check them again
2. if guest keep running, and PID of virsh console guest was gone, re-trigger virsh console guest again
3. if guest was not running, sleep 3 sec to check guest status again 
```
  
Do these console processes constantly got terminated for any reason?
`right now, just make sure reboot guest will terminated  guest console`

Reply:
This reboot log from serial console is the most valuable log here, especially when guest boot fail. In our tests , quite some places have reboot/start operation, with this regard, 3 seconds is reasonable.

Reply:
Thanks for the thoughtful considerations here. Great job!